rfq/oracle.go

```diff
@@ -324,6 +325,42 @@ func (r *RpcPriceOracle) QueryBidPrice(ctx context.Context, assetId *asset.ID,
 	// set the subject asset ID.
 	copy(subjectAssetId, assetId[:])
 
+	// Construct the RPC asset rates hint.
+	var (
+		rpcAssetRatesHint *oraclerpc.AssetRates
+		err               error
+	)
+	assetRateHint.WhenSome(func(assetRate rfqmsg.AssetRate) {
```

Comment:
Should use `fn.MapOptionZ` if there's an error return variable.

Reply:
Even if I use `fn.MapOptionZ` here I'll still need to modify one variable in the outer scope. Using `fn.Result` doesn't seem to work because the "zero" value will be an error at best I think.

Reply:
Here are my two proposals for a different approach (preferring variant 1):

[variant1.diff.txt](https://github.com/user-attachments/files/17562008/variant1.diff.txt)
[variant2.diff.txt](https://github.com/user-attachments/files/17562007/variant2.diff.txt)

Reply:
Unit tests don't pass on variant 1. That's what I ran into also.

When `assetRateHint` is `None`, we want to end up with `rpcAssetRatesHint` with value of `nil`. But instead we end up with `rpcAssetRatesHint` with value of `AssetRates{}`.

And when the price oracle service inspects the `QueryAssetRatesRequest.AssetRatesHint` field it is present but unset instead of just `nil`. In my mind that's an invalid request because the expiry timestamp is 0.

Reply:
In [variant2.diff.txt](https://github.com/user-attachments/files/17562007/variant2.diff.txt) the err is returned but the result is applied to an out of scope variable. I don't see the win there.

Reply:
Okay, I guess variant 1 would then need to have a pointer in the result, then it would properly be `nil`. Not super nice to have a pointer within an option/result, but at least we'd keep the composability of those functional/optional uses and don't go the "side effect" route...